src/main/java/com/hubspot/jackson/datatype/protobuf/internal/PropertyNamingCache.java

```diff
@@ -0,0 +1,100 @@
+import com.hubspot.jackson.datatype.protobuf.ProtobufJacksonConfig;
+
+public class PropertyNamingCache {
+  private final Descriptor descriptor;
+  private final ProtobufJacksonConfig config;
+  private final Map<PropertyNamingStrategy, Function<FieldDescriptor, String>> serializationCache;
+  private final Map<PropertyNamingStrategy, Function<String, FieldDescriptor>> deserializationCache;
+
+  private PropertyNamingCache(Descriptor descriptor, ProtobufJacksonConfig config) {
+    this.descriptor = descriptor;
+    this.config = config;
+    this.serializationCache = Collections.synchronizedMap(new WeakHashMap<>());
+    this.deserializationCache = Collections.synchronizedMap(new WeakHashMap<>());
+  }
+
+  public static PropertyNamingCache forDescriptor(Descriptor descriptor, ProtobufJacksonConfig config) {
+    return new PropertyNamingCache(descriptor, config);
+  }
+
+  public Function<FieldDescriptor, String> forSerialization(PropertyNamingStrategy propertyNamingStrategy) {
+    Function<FieldDescriptor, String> cached = serializationCache.get(propertyNamingStrategy);
+    if (cached != null) {
+      return cached;
+    } else {
+      Function<FieldDescriptor, String> function = buildSerializationFunction(descriptor, propertyNamingStrategy);
+      serializationCache.put(propertyNamingStrategy, function);
+
+      return function;
+    }
+  }
+
+  public Function<String, FieldDescriptor> forDeserialization(PropertyNamingStrategy propertyNamingStrategy) {
+    Function<String, FieldDescriptor> cached = deserializationCache.get(propertyNamingStrategy);
+    if (cached != null) {
+      return cached;
+    } else {
+      Function<String, FieldDescriptor> function = buildDeserializationFunction(descriptor, propertyNamingStrategy);
+      deserializationCache.put(propertyNamingStrategy, function);
+
+      return function;
+    }
+  }
+
+  private Function<FieldDescriptor, String> buildSerializationFunction(
+      Descriptor descriptor,
+      PropertyNamingStrategy originalNamingStrategy
+  ) {
+    PropertyNamingStrategyBase namingStrategy =
+        new PropertyNamingStrategyWrapper(originalNamingStrategy);
+
+    Map<FieldDescriptor, String> tempMap = new HashMap<>();
+    for (FieldDescriptor field : descriptor.getFields()) {
+      tempMap.put(field, namingStrategy.translate(field.getName()));
+    }
+
+    ImmutableMap<FieldDescriptor, String> fieldLookup = ImmutableMap.copyOf(tempMap);
+    return field -> {
+      String name = fieldLookup.get(field);
+      return name == null ? namingStrategy.translate(field.getName()) : name;
+    };
+  }
+
+  private Function<String, FieldDescriptor> buildDeserializationFunction(
+      Descriptor descriptor,
+      PropertyNamingStrategy originalNamingStrategy
+  ) {
+    PropertyNamingStrategyBase namingStrategy =
+        new PropertyNamingStrategyWrapper(originalNamingStrategy);
+
+    Map<String, FieldDescriptor> tempMap = new HashMap<>();
```

Comment:
same comment here around building immutable map native.

Reply:
Later on we use `containsKey` while building the map, which you can't do with an `ImmutableMap.Builder`. And if you insert duplicates, it will throw on construction